may-2018-reenabled-opcodes.md

```diff
@@ -0,0 +1,428 @@
+
+    a b OP_DIV -> out
+    
+    where a and b are interpreted as numeric values
+    
+The operator must fail if:
+1. `!isnum(a) || !isnum(b)`. Fail if either operand is not a numeric value.
+1. `b == 0`. Fail if `b` is equal to any type of zero.
+
+Impact of successful execution:
+* stack memory use reduced
+* number of elements on stack is reduced by one
+
+Unit tests:
+1. `a b OP_DIV -> failure` where `!isnum(a)` or `!isnum(b)`. Both operands must be numeric values
+2. `a 0 OP_DIV -> failure`. Division by positive zero (all sizes), negative zero (all sizes), `OP_0` 
+4. `27 7 OP_DIV -> 3`, `27 -7 OP_DIV -> -3`, `-27 7 OP_DIV -> -3`, `-27 -7 OP_DIV -> 3`. Check negative operands. 
+  *Pay attention to sign*.
+5. check valid results for operands of different lengths `0..4`
+    
+### OP_MOD
+
+    Opcode (decimal): 151
+    Opcode (hex): 0x97
+
+Returns the remainder after dividing a by b.  The output will be represented using the least number of bytes required. 
+
+	a b OP_MOD -> out
+	
+	where a and b are interpreted as numeric values
+	
+The operator must fail if:
+1. `!isnum(a) || !isnum(b)`. Fail if either operand is not a numeric value.
+1. `b == 0`. Fail if `b` is equal to any type of zero.
+
+Impact of successful execution:
+* stack memory use reduced (one element removed)
+* number of elements on stack is reduced by one
+
+Unit tests:
+1. `a b OP_MOD -> failure` where `!isnum(a)` or `!isnum(b)`. Both operands must be numeric values.
+2. `a 0 OP_MOD -> failure`. Division by positive zero (all sizes), negative zero (all sizes), `OP_0` 
+4. `27 7 OP_MOD -> 6`, `27 -7 OP_MOD -> 6`, `-27 7 OP_MOD -> -6`, `-27 -7 OP_MOD -> -6`. Check negative operands. 
+  *Pay attention to sign*.
+5. check valid results for operands of different lengths `0..4`
+
+## New operations
+
+### OP_BIN2NUM
+
+*`OP_BIN2NUM` replaces `OP_LEFT` and uses it's opcode*
+
+    Opcode (decimal): 128
+    Opcode (hex): 0x80
+
+Convert the byte sequence into a numeric value, including minimal encoding.
+
+    `x1 OP_BIN2NUM -> n`
+
+See also `OP_NUM2BIN`.
+
+Notes:
+* if `x1` is any form of zero, including negative zero, then `OP_0` must be the result
+    
+Examples:
+* `0x0000000002 OP_BIN2NUM -> 0x02`
+* `0x800005 OP_BIN2NUM -> 0x85`
+
+The operator must fail if:
+1. the numeric value is out of the range of acceptable numeric values (currently size is limited to 4 bytes)
+
+Unit tests:
+1. `a OP_BIN2NUM -> failure`, when `a` is a byte sequence whose numeric value is too large to fit into the numeric value 
+    type, for both positive and negative values. 
+1. `0x00 OP_BIN2NUM -> OP_0`. Byte sequences, of various lengths, consisting only of zeros should produce an OP_0 (zero 
+    length array). 
+2. `0x00000000000001 OP_BIN2NUM -> 0x01`. A large byte sequence, whose numeric value would fit in the numeric value 
+    type, is a valid operand.
+1. `0x80000000000001 OP_BIN2NUM -> 0x81`. Same as above, for negative values.  
```

Comment:
Some of these tests fail because 0x80000000000001 is longer than the maximum 4-byte integer allowed by numeric opcodes (see CScriptNum).